torch_np/tests/test_reductions.py

```diff
@@ -0,0 +1,617 @@
+        result_with_out = self.func(a, axis=axis, keepdims=keepdims, out=out)
+
+        assert result_with_out is out
+        assert result_with_out.dtype == dtype
+        assert_array_equal(result, result_with_out)
+
+        # TODO: what if result.dtype != out.dtype; does out typecast the result?
+
+        # out of wrong shape (any/out does not broadcast)
+        # np.any(m, out=np.empty_like(m)) raises a ValueError (wrong number
+        # of dimensions.)
+        # pytorch.any emits a warning and resizes the out array.
+        # Here we follow pytorch, since the result is a superset
+        # of the numpy functionality
+
+    @pytest.mark.parametrize('keepdims', [True, False, None])
+    @pytest.mark.parametrize('dtype', [bool, 'int32', 'float64'])
+    def test_out_axis(self, dtype, keepdims):
+        for axis in self.allowed_axes + [None]:
+            self._check_out_axis(axis, dtype, keepdims)
+
+    def _check_keepdims_out(self, axis):
+        """Check the expicit shape of out w/keepdims=True"""
+        d = np.ones((3, 5, 7, 11))
+        if axis is None:
+            shape_out = (1,) * d.ndim
+        else:
+            axis_norm = _util.normalize_axis_tuple(axis, d.ndim)
+            shape_out = tuple(
+                1 if i in axis_norm else d.shape[i] for i in range(d.ndim))
+        out = np.empty(shape_out)
+        result = self.func(d, axis=axis, keepdims=True, out=out)
+        assert result is out
+        assert_equal(result.shape, shape_out)
+
+    def test_keepdims_out(self):
+        allowed_axes = [ax for ax in self.allowed_axes if ax != ()]
+        for axis in allowed_axes + [None,]:
+            self._check_keepdims_out(axis)
+
+
+
+
+class TestAnyGeneric(_GenericReductionsTestMixin, _GenericHasOutTestMixin):
+    def setup_method(self):
+        self.func = np.any
+        self.allowed_axes =  [0, 1, 2, -1, -2,]
+
+    @pytest.mark.xfail(reason='XXX: pytorch does not support any(..., axis=tuple)')
+    def test_axis_bad_tuple(self):
+        super().test_axis_bad_tuple()
+
+
+class TestAllGeneric(_GenericReductionsTestMixin, _GenericHasOutTestMixin):
+    def setup_method(self):
+        self.func = np.all
+        self.allowed_axes =  [0, 1, 2, -1, -2,]
+
+    @pytest.mark.xfail(reason='XXX: pytorch does not support all(..., axis=tuple)')
+    def test_axis_bad_tuple(self):
+        super().test_axis_bad_tuple()
+
+
+class TestCountNonzeroGeneric(_GenericReductionsTestMixin):
+    # count_nonzero does not have the out=... argument
+    def setup_method(self):
+        self.func = np.count_nonzero
+        self.allowed_axes =  [0, 1, 2, -1, -2,
+                              (0, 1), (1, 0), (0, 1, 2), (1, -1, 0)]
+
+
+class TestArgminGeneric(_GenericReductionsTestMixin, _GenericHasOutTestMixin):
+    def setup_method(self):
+        self.func = np.argmin
+        self.allowed_axes =  [0, 1, 2, -1, -2, ]
+
+    @pytest.mark.xfail(reason='XXX: argmin does not allow axis=tuple')
```

Comment:
Why are these tests xfailed? I don't quite understand the reason.

Reply:
The underlying reason is that `arg{min, max}` only accept a single axis and reject axis tuples. This is both numpy and pytorch:

```
In [176]: t
Out[176]: 
tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])

In [177]: t.argmin(dim=(0, 1))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [177], in <cell line: 1>()
----> 1 t.argmin(dim=(0, 1))

TypeError: argmin(): argument 'dim' must be int, not tuple

In [178]: t.numpy().argmin(axis=(0, 1))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [178], in <cell line: 1>()
----> 1 t.numpy().argmin(axis=(0, 1))

TypeError: 'tuple' object cannot be interpreted as an integer
```

Now xfailing these tests is just a technical implement: generic tests inherited from `_GenericReductionTestsMixin` test both integer axis and axis tuples. So it's either split the mixin into two (which adds complexity) or xfail in rare cases. Now that I see that it's not clear for a reader/reviewer : I'll add a comment, would that be enough? Or would `pytest.mark.skip` be clearer?

Reply:
I like `xfail`s for bugs and `skip`s for things that we know that fail and we accept that. That way, if at some point we fix the given bug CI will remind us to remove those xfails. What threw me off was the word `axis` in the xfail comment, as in PyTorch we call it `dim` or `dims`.

About this particular test, I see that both implementations error out on tuples, so I still don't understand what's the difference between the two, sorry :/

Reply:
OK so the xfails are actually not needed: the tests were in fact xpassing. These tests check that basically,
- `reduction_func(a, axis=(1, 1))` errors out on a repeated axis
- `reduction_func(a, axis=())` behaves as expected.

Now both behaviors are handled before handling the heavy lifting to pytorch, so both behaviors are consistent with numpy now (in a couple of minutes when I push).

Reply:
how come you removed the whole test, rather than just hte xfails?

Reply:
Nope, the test is there through the inheritance:

```
class _GenericReductionsTestMixin:        # class name does not start with Test, pytest skips it
    def test_axis_bad_tuple(self):
        # Basic check of functionality
        m = np.array([[0, 1, 7, 0, 0], [3, 0, 0, 2, 19]])
        with assert_raises(ValueError):
             self.func(m, axis=(1, 1))                   # note self.func

class TestAnyGeneric(_GenericReductionsTestMixin):      # class name starts with Test
    def setup_method(self):
        self.func = np.any
```

I see that the whole xfail story was confusing, but it is gone now anyway :-).